### ISSUE TITLE
Fresh install of Shopify CLI gets this extension stuck waiting for CLI version output

### ISSUE THREAD

**Issue by ADTC** [Shopify/theme-check-vscode#42]
_Note: I'm not sure if this affects Windows only, or it affects all systems. Wherever Shopify CLI will ask for usage reporting, it will likely get stuck._

If you freshly installed Shopify CLI but **never** used it using the `shopify` command, then we may likely encounter this issue.

![Task Manager Screenshot](https://user-images.githubusercontent.com/6047296/129631357-2c094ba7-40db-45d1-9be6-a288ff83b60b.png)

The problem is, when this extension executes `shopify version` in the background, it gets stuck waiting for an output. Every time you relaunch the Liquid language server, it creates a new process of that command. _(This can lead to high CPU usage. At one point I had five such processes, each consuming over 20% of CPU - hence my entire CPU was being burned!)_

**Why?** Run `shopify version` by yourself on the command prompt for the first time, and you get a question asking you to enable anonymous usage reporting. Once you have already chosen an answer, the prompt doesn't appear again.

![image](https://user-images.githubusercontent.com/6047296/129632094-50d5882e-dc9a-4a82-a029-734e864856db.png)

After I did this, I restarted VS Code and killed the rogue `shopify version` processes. It seems now that the extension is finally able to get the version number and it tries to start the language server:

![image](https://user-images.githubusercontent.com/6047296/129632882-8bda9ba1-cdfc-4a63-8aa2-531936652dfe.png)

Perhaps there should be some form of fail-safe for the case of fresh install? I'd suggest timing out waiting for version, killing the process then asking the user to run `shopify version` on the command line to check.

**Comment by ADTC** [comment]
_After fixing the above problem:_

It appears the problem of having a rogue process continues to occur. Here's the result when I close and re-open VS Code (same project):

![image](https://user-images.githubusercontent.com/6047296/129759242-0dc29ed2-0930-4882-9313-253f3887c4ef.png)

Only one `cmd.exe` process running the language server.

![image](https://user-images.githubusercontent.com/6047296/129759385-0bf5081f-075c-4fcb-a1f3-2da65a3826f8.png)

Two `ruby.exe` processes running the language server. One is associated with the above `cmd.exe` while the other is rogue. _If we have three more similar rogues here, we'd easily hit 100% CPU usage._

It seems, when I shut down VS Code, the `cmd.exe` process is terminated, but the `ruby.exe` continues to run. Perhaps a solution would be to directly run and manage the `ruby.exe` process on Windows systems, rather than proxying it through `cmd.exe`.

Unless there's a way to have `ruby.exe` go away when `cmd.exe` quits.

**Comment by charlespwd** [comment]
Hey @ADTC :wave: First, thanks a _lot_ for all this information, opening this issue and even trying to debug the issue yourself. Again. 🙇‍♂️ 

I'm trying to isolate whether this is a theme-check-vscode bug, a theme-check bug or a Shopify CLI bug. Would it be too much to ask you to try to do this with the `theme-check-language-server.bat.bat` file set in shopifyLiquid.languageServerPath? Does the issue happen still?

You'd need to install theme-check directly for this and make sure that its version is >= 1.2.0. 

I have a feeling that this is happening because of the `\r\n` -> `\r\r\n` bug in theme-check who'se fix hasn't made it into the Shopify CLI yet.

**Comment by ADTC** [comment]
> Would it be too much to ask you to try to do this with the `theme-check-language-server.bat.bat` file set in shopifyLiquid.languageServerPath?

I have already deleted the bad `.bat` file and renamed `.bat.bat` to `.bat` instead. You can see it below. I have also set the `.bat.bat` file, which is now renamed to `.bat` to the `shopifyLiquid.languageServerPath` setting.

![image](https://user-images.githubusercontent.com/6047296/129766467-f567e999-4fe0-4a3a-b440-acb11c5bce4f.png)

There's are processes running from this:

![image](https://user-images.githubusercontent.com/6047296/129767188-e1696956-b519-4cb9-a7b9-37e3b1889e3c.png)
![image](https://user-images.githubusercontent.com/6047296/129767341-6acd6604-c463-489c-b7a7-a9d5f4534a55.png)

In this case, terminating the manually run `cmd.exe` process will also kill off the corresponding `ruby.exe` process. Not sure why the same doesn't happen for the process run via this extension in VS Code.

I've configured it in `settings.json` in VS Code:

![image](https://user-images.githubusercontent.com/6047296/129767703-13bf71e3-4d17-4c78-a0b0-4b491f8c8123.png)

> You'd need to install theme-check directly for this and make sure that its version is >= 1.2.0.

I'm not sure what you mean. Do you have a command? It was installed as part of Shopify CLI and when I type `shopify theme check -v` it gives the version number as `1.2.0`. So it should be correct?

Still, the language server isn't working. I don't get any object suggestions in VS Code.

**Comment by charlespwd** [comment]
OK. Just found out that I was misunderstanding gem dependencies. `~> 1.1` means greater or equal to `1.1` but less than `2.0`. Somehow, I thought that it only allowed patches (1.1.X) but not minor bumps (1.X). At least now I know that and starts a different perspective on all of this.

1. I understand this one. We should handle this in here or in Shopify CLI. We handle it in [`lighthouse-ci-action`](https://github.com/Shopify/lighthouse-ci-action/blob/main/entrypoint.sh#L119-L123) so maybe do something similar?
2. This is kind of scary. Would need to look into why.
3. So not even errors? Or are you seeing errors/warnings/etc?
4. Looks like it's something that Gem does for us. Shopify CLI doesn't seem to have that problem, so maybe we need to inspire ourselves by what this team does in their gemspec file.
5. This seems like a theme-check bug. Would need to investigate this. Something about not capturing signals the way we'd expect [here](https://github.com/Shopify/theme-check/blob/main/lib/theme_check/language_server/server.rb#L39-L41)? Or a different Windows bug [here](https://github.com/Shopify/theme-check/blob/main/lib/theme_check/language_server/server.rb#L129-L146)?


Bonus answers:
1. This doesn't happen on Linux or Mac. The command runs and it doesn't run long enough to appear in htop.
2. Not happening either. Once the server analyzed the theme, the CPU impact of having the server idle is 0%.
3. You should see something complete when you type `{{ all_<c-space>`. You should be able to cmd-click the path in `{% render "path" %}` and be redirected to snippets/path.liquid. Theme check errors should appear in the problems tab.

https://user-images.githubusercontent.com/4990691/129783421-4e30ef31-0016-476a-920c-e15b8c97a20b.mp4

   * If you set `shopifyLiquid.trace.server = "verbose"`, then in the Outputs > Theme Check Language Server, you should see the Language Server logs of what is happening.


https://user-images.githubusercontent.com/4990691/129783694-d70a0922-0645-4704-9e5a-6f860d086f18.mp4



4. No need for bat files on linux or mac so no weird Gem stuff happening here.
5. Restarting the server from VS Code stops the process. Closing VS Code stops the process.

**Comment by ADTC** [comment]
> 3\. So not even errors? Or are you seeing errors/warnings/etc?

No, nothing. I checked Output pane's _Extensions_ tab and _Log (Extension Host)_ tab.

> set `shopifyLiquid.trace.server = "verbose"`, then in the Outputs > Theme Check Language Server

Looks normal, no errors. First line is `[Trace - 3:05:36 AM] Sending request 'initialize - (0)'.` followed by a JSON dump. That's it. See the output here: [shopify-theme-language-server-verbose.txt](https://github.com/Shopify/theme-check-vscode/files/7002486/shopify-theme-language-server-verbose.txt)


> 3\. You should see something complete when you type `{{ all_<c-space>`. You should be able to cmd-click the path in `{% render "path" %}` and be redirected to snippets/path.liquid. Theme check errors should appear in the problems tab.

None of these happen. All I get is syntax highlighting 😆
![image](https://user-images.githubusercontent.com/6047296/129786606-1e6edb1c-6620-4233-a756-a34f65fc705c.png)

Not a Ruby expert, so I can't help with the code bits. Perhaps the code assumes *nix systems?

**Comment by charlespwd** [comment]
> Looks normal, no errors. First line is [Trace - 3:05:36 AM] Sending request 'initialize - (0)'. followed by a JSON dump. That's it. See the output here: shopify-theme-language-server-verbose.txt

This is the same symptoms of the I _believe_ I fixed in 1.2.0 to account for Windows systems. 

What should happen:
1. VS Code sends an initialize JSONRPC call
2. theme-check-language-server replies with a response and some capabilities
3. VS Code sends an `initialized` JSONRPC call

[example output](https://gist.github.com/18863ec181370f8ded49c2c9c61f1cdf)

What happens:

1. VS Code sends an "inialize" JSONRPC call to the server
2. Nothing is replied, the server loop is never started

So, somehow, either my fix is incorect (very possible!) or we're running on code that doesn't have it. But since you've clearly shown that you're on 1.2.0, I'm not too sure what's going on.

Can you try commenting out the shopifyCLIPath configuration? When present it takes precedence over theme-check-language-server. I just want to be 100% sure that the bug is coming from theme-check :sweat_smile:

**Comment by ADTC** [comment]
> Can you try commenting out the shopifyCLIPath configuration?

Weird. As soon as I did that, in the **first** Theme Check Language Server output pane (where it got stuck waiting for `initialize` response), it sent a `shutdown` request.

Then there's a new **second** Theme Check Language Server output pane. In here, I got what you expected: `initialize`, response, `initialized`, a bunch of documents got checked, got the conclusion: `Found 28 offenses in 0.11s` 😆 

I noticed the command line for the new language server is noticeably different (second line here is the correct one):
![image](https://user-images.githubusercontent.com/6047296/129804886-8c5a7b33-15bf-4505-a3c2-725366e3a1d0.png)
Note, in here the first line is the correct one:
![image](https://user-images.githubusercontent.com/6047296/129804958-1dcc3a73-c5a2-41ee-b4d9-4eb299c71b40.png)
That could be a clue to the problem.

I left the `shopifyLiquid.languageServerPath` enabled BTW. Should I comment that out too?

The extension is already partially working. **Seems like problem lint is failing though.** The `assign x = 1` doesn't show an unused annotation. The rest are working:
![image](https://user-images.githubusercontent.com/6047296/129805755-00332963-1364-44b3-9768-50552ab4e3e3.png)

The shutdown of the server is also working correctly now.

**Comment by ADTC** [comment]
Things don't work everywhere. So far:

- As said, problem linting doesn't seem to work yet.
- `include` doesn't hyperlink the name. (I know, it's deprecated supposedly, but still.)
- Typing `all_` inside a `liquid` tag doesn't show the same suggestions.
- Typing something like `shop.` or `product.` anywhere doesn't give the list of valid attributes.

Maybe this is the reason why problem linting isn't working. I found this in _Log (Extension Host)_:
```
[2021-08-18 05:53:16.695] [exthost] [error] Error: Illegal argument: range
	at D (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:5:1232)
	at new ct (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:78:37553)
	at new ct (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:78:17347)
	at new r (c:\Users\ADTC\.vscode\extensions\shopify.theme-check-vscode-1.1.2\dist\extension.js:1:162568)
	at Y (c:\Users\ADTC\.vscode\extensions\shopify.theme-check-vscode-1.1.2\dist\extension.js:1:156667)
	at Array.map (<anonymous>)
	at asDocumentLinks (c:\Users\ADTC\.vscode\extensions\shopify.theme-check-vscode-1.1.2\dist\extension.js:1:160824)
	at async G.provideLinks (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:85:179647)
[2021-08-18 05:53:17.861] [exthost] [error] [Shopify.theme-check-vscode] provider FAILED
[2021-08-18 05:53:17.861] [exthost] [error] Error: Illegal argument: range
	at D (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:5:1232)
	at new ct (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:78:37553)
	at new ct (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:78:17347)
	at new r (c:\Users\ADTC\.vscode\extensions\shopify.theme-check-vscode-1.1.2\dist\extension.js:1:162568)
	at Y (c:\Users\ADTC\.vscode\extensions\shopify.theme-check-vscode-1.1.2\dist\extension.js:1:156667)
	at Array.map (<anonymous>)
	at asDocumentLinks (c:\Users\ADTC\.vscode\extensions\shopify.theme-check-vscode-1.1.2\dist\extension.js:1:160824)
	at async G.provideLinks (c:\Program Files\Microsoft VS Code\resources\app\out\vs\workbench\services\extensions\node\extensionHostProcess.js:85:179647)
```

**Comment by charlespwd** [comment]
Woo!! Progress! Interesting!

> * include doesn't hyperlink the name. (I know, it's deprecated supposedly, but still.)

We actually have [an issue](https://github.com/Shopify/theme-check/issues/243) for this in theme-check already. It's in our backlog :)

> * Typing all_ inside a liquid tag doesn't show the same suggestions.

You should only get completions if you are in a variable lookup spot. So in a liquid tag, this should complete on the right hand side of an assign, in an if condition, after the `in` in a `for` loop, etc.

Try these:

```
{% liquid
  assign x = all_
  echo all_
  for i in all_
  endfor
  if all_
  endif
%}
```

> * Typing something like shop. or product. anywhere doesn't give the list of valid attributes.

That's also a TODO in theme-check (It's a [WIP](https://github.com/Shopify/theme-check/pull/273) actually)

> Weird. As soon as I did that, in the first Theme Check Language Server output pane (where it got stuck waiting for initialize response), it sent a shutdown request (didn't work though).
> Then there's a new second Theme Check Language Server output pane. In here, I got what you expected: initialize, response, initialized, a bunch of documents got checked, got the conclusion: Found 28 offenses in 0.11s 😆

This actually kind of makes sense. When you restart the server, a new set of logs is created. The first server is closed and a new one starts. The first one didn't close because it never initialized (stuck in an infinite loop somewhere probably). The second one working is really promising! Next time you open VS Code you should only have this one open. 

> I left the shopifyLiquid.languageServerPath enabled BTW. Should I comment that out too?

No! That's what making things work right now :) Turn it off and you'll be using the one from the shopify CLI again.

------

In summary, I think we're now faced with two issues:

1. `shopify theme language-server` doesn't behave like `theme-check-language-server` on Windows even though they are both using the same theme-check version.
2. There seems to be a problem with the document links (the illegal range thingy)

**Comment by ADTC** [comment]
> Try these:

Working! Thanks. Now I understand what's happening. It's smart enough to know the context!

> That's also a TODO in theme-check (It's a [WIP](https://github.com/Shopify/theme-check/pull/273) actually)

Hope to see it released soon. I'm too used to looking up object attributes using `.` — it's muscle memory from a decade of using Java where the IDE feature is a standard and expected.

---

Any idea why problem linting isn't working here? Example: No unused variable annotation.

And is there anything I can help with, to sort out the Windows support?

**Comment by charlespwd** [comment]
> Any idea why problem linting isn't working here? Example: No unused variable annotation.

This is actually surprising. Did you save the file prior? Checks are only run on saved files (because most of them expect valid syntax, otherwise you'd be swamped by SyntaxErrors while typing)

If not, then we should open an issue in theme-check for that. I'm not having this kind of issue:

https://user-images.githubusercontent.com/4990691/129911632-304f8930-13a3-4855-8078-a438a08d46a8.mp4

> And is there anything I can help with, to sort out the Windows support?

We have quite a good lead for debugging this issue now! We'll let you know when we think we have a fix for this :) 

Again, thanks a LOT for your help with this @ADTC !!! 🙇‍♂️

**Comment by ADTC** [comment]
Thanks for [that](https://github.com/Shopify/theme-check/issues/395). I did [one](https://github.com/Shopify/theme-check/issues/396) too! Have a good day ahead. 🙂

**Comment by charlespwd** [comment]
Hi @ADTC, we just did two things:

1. Released a new version of the VS Code extension that no longer does the version check
2. Released theme-check 1.3 that fixes the .bat.bat issue

Are you still still having CPU problems?

**Comment by charlespwd** [comment]
Aye aye aye. We had different bugs in that version. We just released 1.5.0. I tested it this morning from `gem install theme-check` on Windows and it _looks_ like it's fixed.

**Comment by charlespwd** [comment]
I'll close this for now, feel free to reopen if this is still a problem :see_no_evil:
